src/Microsoft.AspNet.Mvc.Core/SessionStateTempDataProvider.cs

```diff
@@ -16,8 +21,31 @@ namespace Microsoft.AspNet.Mvc
     /// </summary>
     public class SessionStateTempDataProvider : ITempDataProvider
     {
-        private static JsonSerializer jsonSerializer = new JsonSerializer();
-        private static string TempDataSessionStateKey = "__ControllerTempData";
+        private const string TempDataSessionStateKey = "__ControllerTempData";
+        private readonly JsonSerializer _jsonSerializer = JsonSerializer.Create(
+            new JsonSerializerSettings()
+            {
+                TypeNameHandling = TypeNameHandling.None
+            });
+
+        private static readonly MethodInfo _convertArrayMethodInfo = typeof(SessionStateTempDataProvider).GetMethod(
+            nameof(ConvertArray), BindingFlags.NonPublic | BindingFlags.Static, null, new[] { typeof(JArray) }, null);
+
+        private readonly ConcurrentDictionary<Type, Func<JArray, object>> _arrayConverters =
```

Comment:
Could you add a note here - https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNet.Mvc/MvcServices.cs#L169 - saying that it needs to be a singleton because we have things that are cached on it?